File: typescript/integrations/langchain-tool.ts
/**
 * Integration: AGIRAILS as LangChain Tool
 *
 * Demonstrates how to wrap AGIRAILS services as LangChain tools,
 * enabling AI agents built with LangChain to pay for and consume
 * services on the AGIRAILS network.
 *
 * This allows any LangChain agent to:
 * - Discover and use paid AI services
 * - Make autonomous payments for work
 * - Integrate with the agent economy
 *
 * Prerequisites:
 *   npm install @langchain/core @langchain/openai zod
 *
 * Run: npx ts-node integrations/langchain-tool.ts
 */

import { DynamicStructuredTool } from '@langchain/core/tools';
// For full agent setup (requires OpenAI API key):
// import { ChatOpenAI } from '@langchain/openai';
// import { AgentExecutor, createOpenAIToolsAgent } from 'langchain/agents';
// import { ChatPromptTemplate } from '@langchain/core/prompts';
import { z } from 'zod';
import { provide, request } from '@agirails/sdk';
import { waitForProvider } from '../src/utils/helpers';

// =========================================================
// PART 1: Create AGIRAILS-powered LangChain Tools
// =========================================================

/**
 * Creates a LangChain tool that wraps an AGIRAILS service.
 * The tool handles payment automatically through the ACTP protocol.
 */
function createAGIRAILSTool(
  serviceName: string,
  description: string,
  schema: z.ZodObject<any>,
  budgetUsd: number
): DynamicStructuredTool {
  return new DynamicStructuredTool({
    name: serviceName.replace(/-/g, '_'), // LangChain prefers underscores
    description: `${description} (Cost: $${budgetUsd} USDC, paid via AGIRAILS)`,
    schema,
    func: async (input) => {
      try {
        const { result } = await request(serviceName, {
          input,
          budget: budgetUsd,
        });
        return JSON.stringify(result, null, 2);
      } catch (error) {
        return `Error calling ${serviceName}: ${error}`;
      }
    },
  });
}

// =========================================================
// PART 2: Example Tools
// =========================================================

// Tool 1: Sentiment Analysis
const sentimentAnalysisTool = createAGIRAILSTool(
  'sentiment-analysis',
  'Analyzes the sentiment of text. Returns positive, neutral, and negative scores.',
  z.object({
    text: z.string().describe('The text to analyze for sentiment'),
  }),
  0.5 // $0.50 per analysis
);

// Tool 2: Code Review
const codeReviewTool = createAGIRAILSTool(
  'code-review',
  'Reviews code for bugs, security issues, and improvements.',
  z.object({
    code: z.string().describe('The code to review'),
    language: z.string().describe('Programming language (e.g., typescript, python)'),
  }),
  2.0 // $2.00 per review
);

// Tool 3: Translation
const translationTool = createAGIRAILSTool(
  'translation',
  'Translates text from one language to another.',
  z.object({
    text: z.string().describe('The text to translate'),
    sourceLanguage: z.string().describe('Source language code (e.g., en, hr, de)'),
    targetLanguage: z.string().describe('Target language code'),
  }),
  0.1 // $0.10 per translation
);

// =========================================================
// PART 3: Start Mock AGIRAILS Providers
// =========================================================

async function startMockProviders() {
  console.log('[Providers] Starting mock AGIRAILS services...\n');

  // Sentiment Analysis Provider
  const sentimentProvider = provide('sentiment-analysis', async (job) => {
    const { text } = job.input as { text: string };
    console.log(`[Sentiment] Analyzing: "${text.substring(0, 50)}..."`);

    // Simulate AI processing
    await new Promise((r) => setTimeout(r, 300));

    return {
      positive: 0.7,
      neutral: 0.2,
      negative: 0.1,
      summary: text.length > 50 ? 'Generally positive sentiment' : 'Neutral',
    };
  });

  // Code Review Provider
  const codeReviewProvider = provide('code-review', async (job) => {
    const { code, language } = job.input as { code: string; language: string };
    console.log(`[CodeReview] Reviewing ${language} code (${code.length} chars)`);

    await new Promise((r) => setTimeout(r, 500));

    return {
      language,
      issues: [
        { severity: 'warning', message: 'Consider adding error handling', line: 3 },
        { severity: 'info', message: 'Could use async/await instead of promises', line: 7 },
      ],
      score: 85,
      summary: 'Code is generally well-structured with minor improvements suggested.',
    };
  });

  // Translation Provider
  const translationProvider = provide('translation', async (job) => {
    const { text, sourceLanguage, targetLanguage } = job.input as {
      text: string;
      sourceLanguage: string;
      targetLanguage: string;
    };
    console.log(`[Translation] ${sourceLanguage} -> ${targetLanguage}: "${text.substring(0, 30)}..."`);

    await new Promise((r) => setTimeout(r, 200));

    // Mock translations
    const translations: Record<string, string> = {
      'Hello, world!': 'Hallo, Welt!',
      'How are you?': 'Wie geht es dir?',
    };

    return {
      original: text,
      translated: translations[text] || `[Translated to ${targetLanguage}]: ${text}`,
      sourceLanguage,
      targetLanguage,
    };
  });

  // Wait for all providers to start
  await Promise.all([
    waitForProvider(sentimentProvider),
    waitForProvider(codeReviewProvider),
    waitForProvider(translationProvider),
  ]);

  console.log('[Providers] All services ready!\n');

  return {
    stop: async () => {
      await sentimentProvider.stop();
      await codeReviewProvider.stop();
      await translationProvider.stop();
    },
  };
}

// =========================================================
// PART 4: Create LangChain Agent with AGIRAILS Tools
// =========================================================

async function createAgentWithAGIRAILSTools() {
  // Note: In real usage, you'd use a real OpenAI key
  // For this demo, we'll simulate the agent behavior
  console.log('[Agent] Creating LangChain agent with AGIRAILS tools...\n');

  const tools = [sentimentAnalysisTool, codeReviewTool, translationTool];

  console.log('Available tools:');
  tools.forEach((tool) => {
    console.log(`  - ${tool.name}: ${tool.description}`);
  });
  console.log('');

  return tools;
}

// =========================================================
// PART 5: Demo - Direct Tool Usage
// =========================================================

async function demoDirectToolUsage(tools: DynamicStructuredTool[]) {
  console.log('='.repeat(60));
  console.log('Demo: Direct Tool Usage (simulating LangChain agent calls)');
  console.log('='.repeat(60));
  console.log('');

  // Simulate agent deciding to analyze sentiment
  console.log('[Agent] I need to analyze customer feedback sentiment...\n');
  const sentimentResult = await tools[0].invoke({
    text: 'This product is amazing! Best purchase I ever made. Highly recommend!',
  });
  console.log('[Agent] Sentiment analysis result:');
  console.log(sentimentResult);
  console.log('');

  // Simulate agent deciding to review code
  console.log('[Agent] Now let me review this code snippet...\n');
  const codeResult = await tools[1].invoke({
    code: `
function fetchData(url) {
  return fetch(url)
    .then(res => res.json())
    .then(data => {
      console.log(data);
      return data;
    });
}`,
    language: 'javascript',
  });
  console.log('[Agent] Code review result:');
  console.log(codeResult);
  console.log('');

  // Simulate agent deciding to translate
  console.log('[Agent] Let me translate this greeting...\n');
  const translationResult = await tools[2].invoke({
    text: 'Hello, world!',
    sourceLanguage: 'en',
    targetLanguage: 'de',
  });
  console.log('[Agent] Translation result:');
  console.log(translationResult);
  console.log('');
}

// =========================================================
// MAIN
// =========================================================

async function main() {
  console.log('='.repeat(60));
  console.log('AGIRAILS Integration: LangChain Tools');
  console.log('='.repeat(60));
  console.log('');
  console.log('This example shows how to wrap AGIRAILS services as LangChain tools.');
  console.log('Each tool automatically handles payment via the ACTP protocol.');
  console.log('');

  // Start mock providers
  const providers = await startMockProviders();

  try {
    // Create agent with AGIRAILS tools
    const tools = await createAgentWithAGIRAILSTools();

    // Demo direct tool usage
    await demoDirectToolUsage(tools);

    console.log('='.repeat(60));
    console.log('Summary');
    console.log('='.repeat(60));
    console.log('');
    console.log('Key Integration Points:');
    console.log('  1. createAGIRAILSTool() wraps any AGIRAILS service as a LangChain tool');
    console.log('  2. Each tool call automatically pays the provider via ACTP');
    console.log('  3. Tools can be used with any LangChain agent (OpenAI, Anthropic, etc.)');
    console.log('  4. Budget is specified per-tool, enabling cost control');
    console.log('');
    console.log('Production Usage:');
    console.log('  - Replace mock providers with real AGIRAILS network services');
    console.log('  - Add proper API key for LangChain model (OpenAI, Anthropic)');
    console.log('  - Create custom tools for your specific use cases');
    console.log('');
  } finally {
    await providers.stop();
  }
}

main().catch(console.error);

File: python/integrations/langchain_tool.py
#!/usr/bin/env python3
"""
Integration: LangChain Tool

Demonstrates wrapping AGIRAILS services as LangChain tools.
Enables AI agents to use AGIRAILS services in their reasoning chains.

Requirements:
    pip install langchain langchain-core

Run: python integrations/langchain_tool.py
"""

import asyncio
import sys
from pathlib import Path
from typing import Any, Optional, Type
from dataclasses import dataclass

sys.path.insert(0, str(Path(__file__).parent.parent))

from src.utils.helpers import clear_mock_state, log, log_section


# Check if LangChain is available
try:
    from langchain.tools import BaseTool
    from langchain.callbacks.manager import CallbackManagerForToolRun
    from pydantic import BaseModel, Field
    HAS_LANGCHAIN = True
except ImportError:
    HAS_LANGCHAIN = False
    print("LangChain not installed. Run: pip install langchain langchain-core")
    print("This example will show the conceptual implementation.\n")


@dataclass
class MockToolResult:
    """Mock result for demonstration without LangChain."""
    output: str


def create_agirails_tool_class():
    """Create AGIRAILS tool class (requires LangChain)."""

    if not HAS_LANGCHAIN:
        return None

    class TranslateInput(BaseModel):
        """Input schema for translation tool."""
        text: str = Field(description="Text to translate")
        target_lang: str = Field(default="de", description="Target language code (e.g., 'de', 'es', 'fr')")

    class AGIRAILSTranslateTool(BaseTool):
        """LangChain tool that uses AGIRAILS translation service."""

        name: str = "agirails_translate"
        description: str = "Translate text to another language using AGIRAILS. Input should be the text to translate."
        args_schema: Type[BaseModel] = TranslateInput

        # AGIRAILS configuration
        provider_address: str = "0x2222222222222222222222222222222222222222"
        budget: float = 1.0
        _client: Any = None
        _loop: Any = None

        def __init__(self, client=None, **kwargs):
            super().__init__(**kwargs)
            self._client = client
            self._loop = asyncio.new_event_loop()

        def _run(
            self,
            text: str,
            target_lang: str = "de",
            run_manager: Optional[CallbackManagerForToolRun] = None,
        ) -> str:
            """Execute the tool synchronously."""
            return self._loop.run_until_complete(
                self._arun(text, target_lang, run_manager)
            )

        async def _arun(
            self,
            text: str,
            target_lang: str = "de",
            run_manager: Optional[CallbackManagerForToolRun] = None,
        ) -> str:
            """Execute the tool asynchronously."""
            from agirails.level0 import request

            result = await request(
                "translate",
                input={"text": text, "targetLang": target_lang},
                budget=self.budget,
                client=self._client,
                provider=self.provider_address,
            )

            output = result.result
            return f"Translated ({output.get('sourceLang', 'en')} -> {target_lang}): {output.get('translated', text)}"

    return AGIRAILSTranslateTool


async def main() -> None:
    log_section("AGIRAILS Integration - LangChain Tool")

    clear_mock_state()

    from agirails import ACTPClient
    from agirails.level0 import provide, request, set_provider_client, start_provider, stop_provider

    REQUESTER = "0x1111111111111111111111111111111111111111"
    PROVIDER = "0x2222222222222222222222222222222222222222"

    # =====================================================
    # Part 1: Setup AGIRAILS Service
    # =====================================================
    log_section("Part 1: Setup AGIRAILS Service")

    # Mock translation
    async def translate_handler(data: dict) -> dict:
        text = data.get("text", "")
        target = data.get("targetLang", "de")
        # Simple mock translation
        translated = f"[{target.upper()}] {text}"
        return {
            "original": text,
            "translated": translated,
            "sourceLang": "en",
            "targetLang": target,
        }

    provide("translate", translate_handler)

    client = await ACTPClient.create(mode="mock", requester_address=REQUESTER)
    await client.mint_tokens(REQUESTER, 100_000_000)
    await client.mint_tokens(PROVIDER, 10_000_000)

    set_provider_client(client, address=PROVIDER)
    await start_provider()

    print("Translation service active")
    print()

    # =====================================================
    # Part 2: Create LangChain Tool
    # =====================================================
    log_section("Part 2: Create LangChain Tool")

    if HAS_LANGCHAIN:
        AGIRAILSTranslateTool = create_agirails_tool_class()
        tool = AGIRAILSTranslateTool(client=client)

        print(f"Tool name: {tool.name}")
        print(f"Tool description: {tool.description}")
        print()

        # Test the tool directly
        log("🔧", "Testing tool directly...")
        result = await tool._arun("Hello world", "de")
        print(f"   Result: {result}")
        print()
    else:
        print("Conceptual tool implementation:")
        print()
        print("```python")
        print("class AGIRAILSTranslateTool(BaseTool):")
        print("    name = 'agirails_translate'")
        print("    description = 'Translate text using AGIRAILS'")
        print("    ")
        print("    async def _arun(self, text: str, target_lang: str) -> str:")
        print("        result = await request(")
        print("            'translate',")
        print("            input={'text': text, 'targetLang': target_lang},")
        print("            budget=1.0")
        print("        )")
        print("        return result.result['translated']")
        print("```")
        print()

    # =====================================================
    # Part 3: LangChain Agent Usage (Conceptual)
    # =====================================================
    log_section("Part 3: LangChain Agent Usage")

    print("Using AGIRAILS tools with LangChain agents:")
    print()
    print("```python")
    print("from langchain.agents import initialize_agent, AgentType")
    print("from langchain.chat_models import ChatOpenAI")
    print()
    print("# Create tools")
    print("translate_tool = AGIRAILSTranslateTool(client=client)")
    print("summarize_tool = AGIRAILSSummarizeTool(client=client)")
    print()
    print("# Create agent with tools")
    print("llm = ChatOpenAI(model='gpt-4')")
    print("agent = initialize_agent(")
    print("    tools=[translate_tool, summarize_tool],")
    print("    llm=llm,")
    print("    agent=AgentType.STRUCTURED_CHAT_ZERO_SHOT_REACT_DESCRIPTION,")
    print(")")
    print()
    print("# Run agent")
    print("result = agent.run(")
    print("    'Translate \"The quick brown fox\" to German, then summarize the result.'")
    print(")")
    print("```")
    print()

    # =====================================================
    # Part 4: Benefits
    # =====================================================
    log_section("Part 4: Benefits")

    print("Why use AGIRAILS with LangChain?")
    print()
    print("1. Paid Services in AI Chains")
    print("   - LangChain agent can pay for premium services")
    print("   - Services like translation, analysis, verification")
    print("   - Automatic budget management")
    print()
    print("2. Decentralized AI Ecosystem")
    print("   - Access services from any AGIRAILS provider")
    print("   - No vendor lock-in")
    print("   - Transparent pricing")
    print()
    print("3. Trust and Verification")
    print("   - Escrow ensures payment after delivery")
    print("   - EAS attestations for proof of work")
    print("   - Dispute resolution if needed")
    print()
    print("4. Composability")
    print("   - Chain multiple AGIRAILS services")
    print("   - Combine with other LangChain tools")
    print("   - Build complex AI workflows")

    # =====================================================
    # Part 5: Multiple Tools Example
    # =====================================================
    log_section("Part 5: Multiple Tools")

    print("Creating multiple AGIRAILS tools:")
    print()
    print("```python")
    print("# Translation tool")
    print("translate = AGIRAILSTool(")
    print("    name='translate',")
    print("    service='translate',")
    print("    description='Translate text to another language',")
    print("    budget=1.0,")
    print(")")
    print()
    print("# Code review tool")
    print("code_review = AGIRAILSTool(")
    print("    name='code_review',")
    print("    service='code-review',")
    print("    description='Review code for security issues',")
    print("    budget=5.0,")
    print(")")
    print()
    print("# Summarization tool")
    print("summarize = AGIRAILSTool(")
    print("    name='summarize',")
    print("    service='summarize',")
    print("    description='Summarize long text',")
    print("    budget=2.0,")
    print(")")
    print()
    print("# Use all tools in agent")
    print("agent = initialize_agent(")
    print("    tools=[translate, code_review, summarize],")
    print("    llm=llm,")
    print(")")
    print("```")

    await stop_provider()
    log("🎉", "LangChain integration demo complete!")


if __name__ == "__main__":
    asyncio.run(main())
